backend/services/stats_service.py

```diff
@@ -479,3 +482,122 @@ def update_project_stats(project_id: int):
+                or_(
+                    TaskHistory.action_text == "MAPPED",
+                    TaskHistory.action_text == "VALIDATED",
+                    TaskHistory.action_text == "INVALIDATED",
+                    TaskHistory.action_text == "BAD_IMAGERY",
+                ),
+            )
+            .filter(
+                and_(
+                    func.DATE(TaskHistory.action_date) >= start_date,
+                    func.DATE(TaskHistory.action_date) <= end_date,
+                )
+            )
+            .group_by(
+                TaskHistory.action_text,
+                "day",
+                TaskHistory.project_id,
+                TaskHistory.task_id,
+            )
+            .order_by("day")
+        )
+
+        if org_id:
+            query = query.join(Project, Project.id == TaskHistory.project_id).filter(
+                Project.organisation_id == org_id
+            )
+        if org_name:
+            organisation = OrganisationService.get_organisation_by_name(org_name)
+            query = query.join(Project, Project.id == TaskHistory.project_id).filter(
+                Project.organisation_id == organisation.id
+            )
+        if campaign_id:
+            query = query.join(
+                campaign_projects,
+                campaign_projects.c.project_id == TaskHistory.project_id,
+            ).filter(campaign_projects.c.campaign_id == campaign_id)
+        if project_id:
+            query = query.filter(TaskHistory.project_id.in_(project_id))
+        if country:
+            # Unnest country column array.
+            sq = Project.query.with_entities(
+                Project.id, func.unnest(Project.country).label("country")
+            ).subquery()
+
+            query = query.filter(sq.c.country.ilike("%{}%".format(country))).filter(
+                TaskHistory.project_id == sq.c.id
+            )
+
+        dates_stats = []
+        task_tracker = {
+            "MAPPED": [],
+            "VALIDATED": [],
+            "INVALIDATED": [],
+            "BAD IMAGERY": [],
+        }
+
+        # r -> (58, 3, 'MAPPED', datetime.date(2021, 1, 5)):
+        # r[0]: task_id, r[1]: project_id,r[2]: action_text, r[3]: date
+        dates = list(set(r[3] for r in query))
+        dates.sort(reverse=False)
+
+        for d in dates:
+            day_stats_dto = TaskStatsDTO(
+                {
+                    "date": d,
+                    "mapped": 0,
+                    "validated": 0,
+                    "invalidated": 0,
+                    "bad_imagery": 0,
+                }
+            )
+
+            values = [(r[0], r[1], r[2]) for r in query if d == r[3]]
+            values.sort(reverse=True)
+
+            for value in values:
+                task_detail = (value[0], value[1])  # tuple for (task_id, project_id)
+                task_status = value[2]
+
+                if task_status == "MAPPED":
+                    if task_detail not in task_tracker["MAPPED"]:
```

Comment:
A mapped task could be validated or invalidated. Also a validated task could be reset to `mapped` state through `Request revalidation` action. If a task has been validated and then it was reset to `mapped` state, then it should be counted only for validation. So we should include an additional check here. Considering the count scenarios, I feel we should have timestamp as a separate field and go through the list in order of the timestamp

Reply:
I tested requesting revalidation in one task, and it worked well. Only the first mapped date was considered.